### ISSUE TITLE
"SaltRenderError: Jinja syntax error: tag name expected" when trying to render Master config file

### ISSUE THREAD

**Issue by dkpiesak** [saltstack/salt#51596]
### Description of Issue/Question
It seems that when pushing the default Master configuration file(as it comes with the installation) to the salt-master instance, it fails with rendering errors. The specific section from the Master config file is the below:

**salt://linux/debian-9/salt_master_config**:
```
....
....
....
# Default Jinja environment options for all templates except sls templates
#jinja_env:
#  block_start_string: '{%'            <==========
#  block_end_string: '%}'             <==========
#  variable_start_string: '{{'
#  variable_end_string: '}}'
#  comment_start_string: '{#'
#  comment_end_string: '#}'
#  line_statement_prefix: 
#  line_comment_prefix: 
#  trim_blocks: False
#  lstrip_blocks: False
#  newline_sequence: '\n'
#  keep_trailing_newline: False

# Jinja environment options for sls templates
#jinja_sls_env:
#  block_start_string: '{%'                <============
#  block_end_string: '%}'                 <============
#  variable_start_string: '{{'
#  variable_end_string: '}}'
#  comment_start_string: '{#'
#  comment_end_string: '#}'
#  line_statement_prefix: 
#  line_comment_prefix: 
#  trim_blocks: False
#  lstrip_blocks: False
#  newline_sequence: '\n'
#  keep_trailing_newline: False
....
....
....
```

The error i get is:

```
salt_master_config
    Function: file.managed
        Name: /etc/salt/master
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1913, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1898, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/states/file.py", line 2545, in managed
                  **kwargs
                File "/usr/lib/python2.7/dist-packages/salt/modules/file.py", line 4739, in check_managed_changes
                  **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/modules/file.py", line 4184, in get_managed
                  **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 170, in render_tmpl
                  output = render_str(tmplstr, context, tmplpath)
                File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 413, in render_jinja_tmpl
                  tmplstr)
              SaltRenderError: Jinja syntax error: tag name expected; line 595

              ---
              [...]
              # The renderer to use on the minions to render the state data
              #renderer: yaml_jinja

              # Default Jinja environment options for all templates except sls templates
              #jinja_env:
              #  block_start_string: '{%'    <======================
              #  block_end_string: '%}'
              #  variable_start_string: '{{'
              #  variable_end_string: '}}'
              #  comment_start_string: '{#'
              #  comment_end_string: '#}'
              [...]
              ---
     Started: 12:28:15.982207
    Duration: 26.908 ms
     Changes:
```

It seems like the rendering of the file fails because there's Jinja characters in that section.
If i escape the `%`, `{` or `}`, it will work fine. It does not make sense why it would fail, cause that entire section is commented out as it ships with the installation.

### Setup
My state looks like:

```
salt_master_config:
  file.managed:
    - name: /etc/salt/master
    - source: salt://linux/debian-9/salt_master_config
    - user: root
    - group: root
    - mode: 640
    - template: jinja
```

### Steps to Reproduce Issue
`salt -v 'target_minion' state.highstate test=True`

### Versions Report
salt-master:
```
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Sep 26 2018, 18:42:22)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.7
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-3-amd64
         system: Linux
        version: debian 9.7
```

salt-minion:
```
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Sep 26 2018, 18:42:22)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.7
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-3-amd64
         system: Linux
        version: debian 9.7
```

Thanks,

**Comment by dwoz** [comment]
@dmurphy18 Can you confirm this is an issue?

**Comment by dkpiesak** [comment]
I forgot to mention that the escape characters that are added in order to make the `state.highstate` work, are pushed to the `master` config file. So, nothing gets actually escaped. 
After the `state.highstate` has run:

```
#jinja_env:
#  block_start_string: '\{\%'          <===
#  block_end_string: '\%\}'           <===
#  variable_start_string: '
......
.....

# Jinja environment options for sls templates
#jinja_sls_env:
#  block_start_string: '\{\%'       <===
#  block_end_string: '\%\}'        <===
#  variable_start_string: '
.....
.....
```

Perhaps i needed to escape the escape characters and create an escape hell there. Didn't try that.
But as long as these lines remain commented out, it does not cause any problem with the Salt Master service starting up.

**Comment by piersf** [comment]
It seems that this version of Saltstack has a real issue with comments that include special characters. I ran into another use case today, where i had the below comment line in pillar/tools/salt-minion/internal_minion.sls:
``` 
### and if '['salt_master_addr']['ip'] doesnot exist in pillars, it breaks the rendering for internal minions 
```

Upon `state.highstate test=True`, it was failing rendering and throwing the below error:
```
2019-02-18 13:06:05,748 [salt.pillar      :741 ][CRITICAL][12228] Rendering SLS 'tools.salt-minion.internal_minion' failed, render error:
Jinja variable u'salt.pillar object' has no attribute 'salt_master_addr'
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 736, in render_pstate
    **defaults)
  File "/usr/lib/python2.7/dist-packages/salt/template.py", line 93, in compile_template
    ret = render(input_data, saltenv, sls, **render_kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/renderers/jinja.py", line 70, in render
    **kws)
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 170, in render_tmpl
    output = render_str(tmplstr, context, tmplpath)
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 403, in render_jinja_tmpl
    buf=tmplstr)
SaltRenderError: Jinja variable u'salt.pillar object' has no attribute 'salt_master_addr'
2019-02-18 13:06:05,900 [salt.pillar      :1026][CRITICAL][12228] Pillar render error: Rendering SLS 'tools.salt-minion.internal_minion' failed. Please see master log for details.
```

Once i removed completely the commented line, it worked without an issue.

**Comment by dmurphy18** [comment]
@dkpiesak @dwoz Yes, that looks like an issue with the use of special Jinja characters in config files, and this should not just be limited Debian 9 but any OS platform where an attempt is made to process a config file, not just master conf.

Need to check all files which could be run through Jinja, similar to using file.managed and ensure that all comments are sufficiently commented out for the render.  In this case for Jinja, we could use '{# ..... #} to isolate the example for Jinja comment processing.  However we should also check for other common rendering, for example: Mako

**Comment by waynew** [comment]
**tl;dr** There may be some other thoughts here, but from my perspective this is working as intended, and is just a (possible) documentation bug.

---

The Salt config files aren't supposed to be Jinja - just YAML, so comment lines can contain any old data, including invalid Jinja, XML, Mako, Moustache... I mean, anything. Cause it's a YAML file, and comments are ignored.

The problem of course that you ran into is that you want to manage your master config file within Salt ([yo dawg](https://memegenerator.net/instance/59388192/xzibit-yo-dawg-yo-dawg-i-heard-you-like-salt-so-i-salted-your-salt-to-make-it-salty)), which is perfectly reasonable.

And you want to stick Jinja in it, which is *also* reasonable.

The problem is that our comments contain characters  that break Jinja templating.

But I don't think that it's necessary (or reasonable) for Salt's default/example config file to avoid breaking templating engines. In fact, your config file could look like this:

```
```

And Salt would happily run with all the defaults. But that wouldn't be useful to anyone.

I think that it *would* be a good idea to make a note somewhere, perhaps on the master config page, that Salt's default config isn't intended to be fed through a template renderer of any kind, and it's expected that you should modify your config files if you want to feed it through a templating engine.

That's my thoughts on it anyway.

**Comment by dmurphy18** [comment]
@dkpiesak Working as intended, the config files are correct since they are yaml files and as such parse correctly with yaml.  There is no intention for the default master config to be deployed via salt as a jinja template.

The documentation shall be updated to clarify the config files intended purpose with yaml.

If this explanation is satisfactory, please consider closing this issue.

**Comment by piersf** [comment]
@dmurphy18 and @waynew thanks for replying guys.

If you check the very first comment above, you'll see that the error is happening even with the default configuration file as it is when `salt-master` is installed, without doing any changes at all.

Furthermore, managing the `salt-master`'s configuration file with Salt has worked just fine in previous versions where we also had Jinja templating inside. In previous versions, I've had tons of Jinja inside the configuration file where I was applying some logic, and that always worked fine. Unless things have changed from the `2018.3.3` version.

**Comment by waynew** [comment]
@piersf So you're saying that this file:

    - source: salt://linux/debian-9/salt_master_config

Only contains valid Jinja?

If that's the case then there definitely is a bug here.

**Comment by dmurphy18** [comment]
@piersf @waynew From your sample state file shows using a copy salt master config as source. so it would appear to be using a yaml file for jinja templating which the config file was not designed to do.

Given the large number of templating solutions, would be difficult to cover them all.
The block_start_string, etc were added two years ago, so it is not a recent change.

I will check previous versions of Salt on Debian 9 tomorrow and see if the issue is due to something else having changed and now catching on this in the config file

**Comment by piersf** [comment]
@waynew yes it contains only valid Jinja - for the most part, the file is the one coming with the installation plus few Jinja lines in the Syndic configuration section where I check the grains['id'] of the host to determine whether it should run Salt-Syndic. Would you like me to paste the entire content of `- source: salt://linux/debian-9/salt_master_config`? I know it's long.

**Comment by waynew** [comment]
@piersf From the original post, unless I'm misunderstanding, you have (truncated):

```
# Default Jinja environment options for all templates except sls templates
#jinja_env:
#  block_start_string: '{%'            <==========
#  block_end_string: '%}'             <==========
#  variable_start_string: '{{'
#  variable_end_string: '}}'
#  comment_start_string: '{#'
```

If that understanding is correct, that is *not* valid Jinja.

To test my hypothesis, change your `salt://linux/debian-9/salt_master_config` file to be an empty file (back up your existing one of course, if you don't have it under version control).

If you have settings like `id: super_cool_msater`, feel free to leave those in.

In fact, I'm pretty sure that, using vim if you run `:g/^#/d` and then save your config file you will not have a problem.

My understanding of what is happening here:

1. Salt comes with a default config file.
2. This file is a *valid* config file. That is, it contains valid YAML
3. In the valid YAML comments (which is a line starting with #, followed by literally anything that isn't a newline), we have invalid Jinja. This is totally fine because Salt doesn't process the config files as anything but YAML (as far as I'm aware)
4. This default config file is being used in your `salt://linux/debian-9/salt_master_config`, with or without some modifications to the relevant sections.
5. Your state ```salt_master_config:
  file.managed:
    - name: /etc/salt/master
    - source: salt://linux/debian-9/salt_master_config
    - user: root
    - group: root
    - mode: 640
    - template: jinja <==== the offending line``` specifies that the master config is a Jinja template
6. Given 3, the assumption in 5 is incorrect.
7. When you run `state.apply` (or highstate), Salt attempts to render the invalid Jinja, and fails.
8. Removing the valid *YAML* comments, containing **invalid Jinja** will solve your problem.

Does this seem accurate to you, or am I missing something?
HTH!

**Comment by dmurphy18** [comment]
@piersf Tested this with Debian 9.0 and Salt 2018.3.1 and it fails as would be expected since yaml file, not a jinja file.  Used the master config file from Salt 2018.3.1, on a clean system.

testcfg.sls
testit:
  file.managed:
    - name: /etc/salt/master
    - source: salt://linux/debian-9/salt_master_config
    - user: root
    - group: root
    - mode: 640
    - template: jinja
oot@deb9min:/home/david# salt d9 test.versions
d9:
    Salt Version:
               Salt: 2018.3.1

    Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 2.5.3
          docker-py: Not Installed
              gitdb: 2.0.0
          gitpython: 2.1.1
              ioflo: Not Installed
             Jinja2: 2.9.4
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: Not Installed
       msgpack-pure: Not Installed
     msgpack-python: 0.4.8
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.13 (default, Jan 19 2017, 14:48:08)
       python-gnupg: Not Installed
             PyYAML: 3.12
              PyZMQ: 16.0.2
               RAET: Not Installed
              smmap: 2.0.1
            timelib: Not Installed
            Tornado: 4.4.3
                ZMQ: 4.2.1

    System Versions:
               dist: debian 9.0
             locale: UTF-8
            machine: x86_64
            release: 4.9.0-3-amd64
             system: Linux
            version: debian 9.0

root@deb9min:/home/david# lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 9.0 (stretch)
Release:	9.0
Codename:	stretch
root@deb9min:/home/david#
root@deb9min:/srv/salt# salt d9 state.sls testcfg test=True
d9:
----------
          ID: testit
    Function: file.managed
        Name: /etc/salt/master
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1905, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 1830, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/states/file.py", line 2396, in managed
                  **kwargs
                File "/usr/lib/python2.7/dist-packages/salt/modules/file.py", line 4726, in check_managed_changes
                  **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/modules/file.py", line 4178, in get_managed
                  **kwargs)
                File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 170, in render_tmpl
                  output = render_str(tmplstr, context, tmplpath)
                File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 409, in render_jinja_tmpl
                  tmplstr)
              SaltRenderError: Jinja syntax error: tag name expected; line 588

              ---
              [...]
              # The renderer to use on the minions to render the state data
              #renderer: yaml_jinja

              # Default Jinja environment options for all templates except sls templates
              #jinja_env:
              #  block_start_string: '{%'    <======================
              #  block_end_string: '%}'
              #  variable_start_string: '{{'
              #  variable_end_string: '}}'
              #  comment_start_string: '{#'
              #  comment_end_string: '#}'
              [...]
              ---
     Started: 11:41:31.747492
    Duration: 59.738 ms
     Changes:

Summary for d9
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:  59.738 ms
ERROR: Minions returned with non-zero exit code

Your statement _Furthermore, managing the salt-master's configuration file with Salt has worked just fine in previous versions where we also had Jinja templating inside._ is not found to be accurate, perhaps some other changes in your environment have changed causing your errors, but it is operating as expected on a clean system.

I would like to close this issue, unless there is some other information you can provide to show that Salt is not working as expected.

**Comment by piersf** [comment]
@dmurphy18 sure, the issue can be closed. 
I will remove the `- template: jinja` from my state. so it is treated as a YAML file and not Jinja template.

@waynew , @dmurphy18  thank you very much for looking into this!
